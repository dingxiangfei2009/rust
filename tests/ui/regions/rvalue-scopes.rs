// edition:2021
// run-pass
// check-run-results
use std::cell::Cell;
use std::ops::{Index, IndexMut};

struct Droppy(u8);
impl Drop for Droppy {
    fn drop(&mut self) {
        println!("AH! {}", self.0)
    }
}

struct A<'a>(Cell<&'a mut Droppy>);

impl<'a> Index<&'a mut Droppy> for A<'a> {
    type Output = Droppy;
    fn index(&self, idx: &'a mut Droppy) -> &Self::Output {
        &*self.0.replace(idx)
    }
}

impl<'a> IndexMut<&'a mut Droppy> for A<'a> {
    fn index_mut(&mut self, idx: &'a mut Droppy) -> &mut Self::Output {
        self.0.replace(idx)
    }
}

fn main() {
    let t = &mut Droppy(0);
    let mut a = A(Cell::new(t));
    a[&mut Droppy(1)] = Droppy(2);
}
