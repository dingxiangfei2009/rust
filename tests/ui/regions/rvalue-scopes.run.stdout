AH! 0
AH! 1
AH! 2
